psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -438,18 +447,28 @@ void DirectJK::compute_JK() {
 }
 void DirectJK::postiterations() {}
 
-void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::vector<std::shared_ptr<Matrix>>& D,
-                        std::vector<std::shared_ptr<Matrix>>& J, std::vector<std::shared_ptr<Matrix>>& K) {
+void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::vector<SharedMatrix>& D,
```

Comment:
I want to be _very clear_ on what this function is doing. In the old code, even when you didn't need both J and K, it would compute both but just write whichever you didn't need to temporary matrices, for reasons that are beyond me. First, is that right?

What does the new function do?

Reply:
Yes, in the old version, temporary buffers are written into, which is a waste of allocated memory, but not significant until you get to large systems. Now, we are running those large systems, and therefore we DO NOT need to allocate unnecessary space if we are building J without K.

The new function allows building J without allocating K buffers (and vice versa), saving time and space if we are using separate J/K builds. Also, if a matrix is already partially built from another method (e.g. J from linK), we have an option to not zero out that matrix.